File: src/renderer/services/backup/formats/version1.ts
import { z } from 'zod'
import { DocHeader } from './version0'
import type { Import } from './version4'

/**
 * Layout v1, will be used in Export v1 but
 * may be reused by later versions.
 */
export type Layouts = z.output<typeof Layouts>
export const Layouts = z.object({
  sources: z.array(
    DocHeader.extend({
      title: z.string().min(1),
      image: z.string().min(1).nullable()
    })
  ),
  switches: z.array(
    DocHeader.extend({
      driverId: z.string().uuid(),
      title: z.string(),
      path: z.string()
    })
  ),
  ties: z.array(
    DocHeader.extend({
      sourceId: z.string().uuid(),
      switchId: z.string().uuid(),
      inputChannel: z.number().int(),
      outputChannels: z
        .object({
          video: z.number().int().optional(),
          audio: z.number().int().optional()
        })
        .default({})
    })
  )
})

/**
 * Backup parser for v1.
 */
export const Backup = z
  .object({
    version: z.literal(1),
    ...Layouts.shape
  })
  .transform(
    (payload): Import => ({
      version: 'normalized' as const,
      settings: {
        buttonOrder: payload.sources.map((source) => source._id)
      },
      layouts: {
        sources: payload.sources.map((source) => ({ order: 0, ...source })),
        devices: payload.switches,
        ties: payload.ties.map(({ switchId, outputChannels, ...tie }) => ({
          ...tie,
          deviceId: switchId,
          outputChannels: {
            video: outputChannels.video ?? undefined,
            audio: outputChannels.audio ?? undefined
          }
        }))
      }
    })
  )

File: composeApp/src/jvmMain/kotlin/org/sleepingcats/bridgecmdr/common/backup/model/Version1.kt
@file:OptIn(ExperimentalUuidApi::class)

package org.sleepingcats.bridgecmdr.common.backup.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlin.uuid.ExperimentalUuidApi
import kotlin.uuid.Uuid

object Version1 : BackupParser<Version5.Backup> {
  @Serializable
  data class Source(
    @SerialName("_id")
    val id: Uuid,
    val title: String,
    val image: String?,
  )

  @Serializable
  data class Switch(
    @SerialName("_id")
    val id: Uuid,
    val driverId: Uuid,
    val title: String,
    val path: String,
  )

  @Serializable
  data class Tie(
    @SerialName("_id")
    val id: Uuid,
    val sourceId: Uuid,
    val switchId: Uuid,
    val inputChannel: Int,
    val outputChannels: OutputChannels = OutputChannels(),
  ) {
    @Serializable
    data class OutputChannels(
      val video: Int? = null,
      val audio: Int? = null,
    )
  }

  @Serializable
  data class Backup(
    val version: Int = 1,
    val sources: List<Source>,
    val switches: List<Switch>,
    val ties: List<Tie>,
  )

  fun upgradeBackup(backup: Backup) =
    Version5.Backup(
      version = 5,
      settings =
        Version5.Settings(
          buttonOrder = backup.sources.map { it.id },
        ),
      layouts =
        Version4.Layouts(
          sources =
            backup.sources.map {
              Version4.Source(
                id = it.id,
                title = it.title,
                image = it.image,
              )
            },
          devices =
            backup.switches.map {
              Version4.Device(
                id = it.id,
                driverId = it.driverId,
                title = it.title,
                path = it.path,
              )
            },
          ties =
            backup.ties.map {
              Version4.Tie(
                id = it.id,
                sourceId = it.sourceId,
                deviceId = it.switchId,
                inputChannel = it.inputChannel,
                outputVideoChannel = it.outputChannels.video,
                outputAudioChannel = it.outputChannels.audio,
              )
            },
        ),
    )

  override fun parseBackup(raw: String) =
    upgradeBackup(
      Json.decodeFromString<Backup>(raw).also { check(it.version == 1) { "Wrong backup version, expected 1" } },
    )
}
